feat: created gilpinflowssystem and demo

lmclane04 opened the pull request:
Added a new system class in src/dynadojo/systems for all of Gilpin's continuous systems (imported from dysts.flows).

Added a demo in demos/systems to test those systems.

lmclane23 committed:
feat: created gilpinflowssystem and demo

Added a new system class in src/dynadojo/systems for all of Gilpin's continuous systems (imported from dysts.flows).

Added a demo in demos/systems to test those systems.

lmclane04 committed:
fix: improve OOD initial condition generation

- Identify principal components that cover up to 90% variance
- Use the remaining principal components (if j = d, then choose j = d-1) for OOD generation

lmclane04 committed:
fix: add docstrings to gilpin flows system

lmclane04 committed:
Merge branch 'main' into feat/gilpin-flows-system/lauren

mkanwal commented on src/dynadojo/systems/gilpin_flows.py:
```diff
@@ -0,0 +1,154 @@
+            x0 = trajectories[:, :, 0]
+            return x0
+        
+        x = self.system.make_trajectory(10000, resample = True)
```
add comments as we discussed

mkanwal commented on src/dynadojo/systems/gilpin_flows.py:
```diff
@@ -0,0 +1,154 @@
+            projection = U_remaining @ random_projection
+            ood_point = mean + scale * projection[:mean.shape[0]]
+            ood_points.append(ood_point)
+
```
stress-test this in the demo notebook across all systems.  visualize the two sets of initial conditions (to check your implementation) and trajectories (to see how well this idea works).

mkanwal commented on src/dynadojo/systems/gilpin_flows.py:
```diff
@@ -0,0 +1,154 @@
+        for i in range(n):
+            self.system.ic = init_conds[i]
+            trajectory = self.system.make_trajectory(timesteps, resample=True)
+            if trajectory.shape[0] < timesteps:
```
add comment

mkanwal commented on src/dynadojo/systems/gilpin_flows.py:
```diff
@@ -0,0 +1,154 @@
+
+        return np.array(ood_points)
+
+    def make_data(self, init_conds: np.ndarray, timesteps: int, control=None, noisy=False):
```
does your latest version include setting time steps per period?

carynbear commented on src/dynadojo/systems/gilpin_flows.py:
```diff
@@ -0,0 +1,154 @@
+import warnings
+import os
+import json
+import dysts
```
This dependency is missing in `pyproject.toml`

Please add a new optional dependencies section called "gilpin" and include any new dependencies. Then add the "gilpin" group to the "all" group. You should commit the pyproject.toml, but not your lockfile.

carynbear commented:
Can you add your system to [`tests/test_deterministic.py`](https://github.com/DynaDojo/dynadojo/blob/main/tests/test_deterministic.py) all you need to do is import and add to the systems list. Then run the tests (see contributing.md) to make sure it passes.

carynbear reviewed (changes requested):
Can you add your system to [`tests/test_deterministic.py`](https://github.com/DynaDojo/dynadojo/blob/main/tests/test_deterministic.py) all you need to do is import and add to the systems list. Then run the tests (see contributing.md) to make sure it passes.

Also fix any comments from @mkanwal and me?

sameerashahh committed:
bug: fit() result dict changed train_losses to train_loss, and val_losses to val_loss (#110)

Vigithai committed:
feat/demo: algorithms/cnn from dmd_demo as template (#93)

oreo07-cyber committed:
refactor: experiments params to config (#118)

* changed all params.json and params_file to config.json and config_file

* changed all params.json and params_file to config.json and config_file. Made experiments config file and ran experiment. All works

* changed last thing from save_params -> save_config

---------

Co-authored-by: Caryn Tran <caryn.tran@berkeley.edu>

lmclane04 committed:
feat: add system/lorenz demo (#88)

Resolves #67

Co-authored-by: lmclane23 <101528096+lmclane23@users.noreply.github.com>

VictorHuynh committed:
feat: add system/lds demo to resolve #66 (#92)

lmclane04 committed:
fix: adjusted GilpinFlowsSystem to produce more varied OOD initial conditions

Changed gilpin_flows demo to plot OOD initial points and also to test OOD points generation for all systems

lmclane04 committed:
fix: address pull request changes

added GilpinFlowsSystem to test_deterministic file
added dysts as optional gilpin system dependency in pyproject.toml
added comments to gilpin_flows.py
added class method to search for all available gilpin systems

lmclane04 committed:
Merge branch 'main' into feat/gilpin-flows-system/lauren

lmclane04 committed:
fix: add gilpin to pyproject.toml

mkanwal commented on pyproject.toml:
```diff
@@ -53,7 +53,7 @@ dev = [
 [project.optional-dependencies]
 all = [
-    "dynadojo[rebound, tensorflow]",
+    "dynadojo[rebound, tensorflow]", gilpin,
```
[rebound, tensorflow, gilpin]",

mkanwal commented on src/dynadojo/systems/gilpin_flows.py:
```diff
@@ -0,0 +1,172 @@
+        
+        return all_systems
+
+    def __init__(self, latent_dim, embed_dim, system_name: str, pts_per_period=100, seed=None):
```
Arguments should all be typed (like you did for system_name). See other systems for example.

Also need to give all params a default value.

mkanwal commented on src/dynadojo/systems/gilpin_flows.py:
```diff
@@ -0,0 +1,172 @@
+
+        mean = np.mean(x, axis=0)
+        x_centered = x - mean
+        U, s, _ = np.linalg.svd(x_centered, full_matrices=False)
```
need to verify this in person with you

lmclane04 committed:
fix: fix dependencies for gilpin systems

also adjusted GIlpinFlowsSystem parameters to take on a default value

lmclane04 committed:
fix: fixes gilpin flows demo

Fixes order of parameters in gilpin flows demo.
Adds warning messages for wrong inputted dimensions.
Adds function return type for make_data.

lmclane04 committed:
fix: generate OOD points along right singular values (not left)

carynbear committed:
add lockfile

carynbear committed:
make deterministic

mkanwal committed:
removed logging

mkanwal committed:
removed logging

lmclane04 committed:
fix: generate more varied initial conditions

generate initial conditions along the sample trajectory.
Perturb points by an amount proportional to the mean and variance of the sample trajectory.

lmclane04 committed:
Merge branch 'main' into feat/gilpin-flows-system/lauren

mkanwal committed:
WIP verifying reproducibility

mkanwal committed:
reverting dependencies
